client/extensions/woocommerce/app/settings/email/mailchimp/setup-mailchimp.js

```diff
@@ -0,0 +1,329 @@
+			if ( active_tab === STORE_SYNC ) {
+				nextProps.onClose();
+			}
+		}
+
+		// Update settings if we have received lists
+		if ( nextProps.settings.mailchimp_lists && ! this.props.settings.mailchimp_lists ) {
+			const settings = this.prepareDefaultValues( nextProps );
+			this.setState( { step: active_tab, settings } );
+		}
+	}
+
+	prepareDefaultValues( nextProps ) {
+		const { settings } = nextProps;
+		const newSettings = Object.assign( {}, settings );
+		newSettings.campaign_from_name = settings.campaign_from_name || settings.store_name || '';
+		newSettings.campaign_from_email = settings.campaign_from_email || settings.admin_email || '';
+		newSettings.campaign_subject = settings.campaign_subject || settings.store_name || '';
+		newSettings.store_locale = settings.store_locale || 'en';
+		newSettings.campaign_language = settings.campaign_language || settings.store_locale;
+		newSettings.campaign_permission_reminder = settings.campaign_permission_reminder ||
+			this.props.translate( 'You were subscribed to the newsletter from ' ) + settings.store_name;
+		newSettings.admin_email = settings.admin_email || nextProps.currentUserEmail || '';
+		newSettings.store_timezone = settings.store_timezone || nextProps.timezone || 'America/New_York';
+		newSettings.mailchimp_lists = settings.mailchimp_lists;
+		newSettings.mailchimp_list = settings.mailchimp_list;
+		return newSettings;
+	}
+
+	getStoreSettings = () => {
+		// clear this and pass only what is required.
+		const { address, currency } = this.props;
+
+		const settings = pick( this.state.settings, storeSettingsRequiredFields );
+		settings.store_city = address.city;
+		settings.store_street = address.street + ' ' + address.street2;
+		settings.store_state = address.state;
+		settings.store_country = address.country;
+		settings.store_postal_code = address.postcode;
+		settings.store_currency_code = currency;
+		return settings;
+	}
+
+	getCampaingDefaultsSettings = () => {
+		return pick( this.state.settings, campaignDefaultsRequiredFields );
+	}
+
+	hasEmptyValues = ( data ) => {
+		return some( data, isEmpty );
+	}
+
+	areStoreSettingsValid = ( settings ) => {
+		const hasAllKeys = storeSettingsRequiredFields.every( key => key in settings );
+		if ( ! hasAllKeys ) {
+			return false;
+		}
+		if ( this.hasEmptyValues( settings ) ) {
+			return false;
+		}
+		if ( settings.store_phone.length <= 6 ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	areCampaignSettingsValid = ( settings ) => {
+		const hasAllKeys = campaignDefaultsRequiredFields.every( key => key in settings );
+		if ( ! hasAllKeys ) {
+			return false;
+		}
+		if ( this.hasEmptyValues( settings ) ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	next = () => {
+		const { step } = this.state;
+		const { siteId } = this.props;
+
+		if ( step === LOG_INTO_MAILCHIMP_STEP ) {
+			this.setState( { step: steps[ this.state.step ].nextStep } );
+		} else if ( step === KEY_INPUT_STEP ) {
```

Comment:
Although I know that state setting occurs asynchronously, you're counting on that here by not returning early inside these conditions.

Would you consider if ( XXXXXX === step) { yyyy;zzzz;return } instead?

Also, please consider yoda-ing

Reply:
if-else-if construction will not be affected by in-flight state change so I believe that return is unnecessary. I can do the Yoda thing though :)